### PR TITLE
Support configurable taggers for query metrics tagging based on context.

### PR DESCRIPTION
**Pull request by vjkoskela**

### CODE REVIEW

File: src/main/java/org/kairosdb/core/reporting/MetricReportingModule.java [hunk 1]
```diff
@@ -76,6 +80,48 @@ public MetricReportingModule(final Properties properties) {
...
+        for (final Map.Entry<String, Class<?>> entry : taggerBuilders.entrySet()) {
+            final String taggerKey = entry.getKey();
+            final Class<?> builderClass = entry.getValue();
+            try {
+                final Object builder = PROPERTIES_MAPPER.readPropertiesAs(
+                        properties,
+                        schema.withPrefix(TAGGER_KEY_PREFIX + taggerKey),
+                        builderClass);
+
+                @SuppressWarnings("unchecked") final org.kairosdb.core.reporting.Tagger tagger =
```
Comment (cwbriones) [inline src/main/java/org/kairosdb/core/reporting/MetricReportingModule.java:115]: Do you need the fully qualified class name here?

File: src/main/java/org/kairosdb/core/reporting/TagTagger.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+         */
+        public Builder setMappedTags(final ImmutableMap<String, String> value) {
+            mappedTags = value;
+            return this;
+        }
+
+        private boolean validate() {
+            final Set<String> mappedTargetTags = Sets.newHashSet(mappedTags.values());
+
+            // Assert that no two mapped dimensions target the same value
```
Comment (cwbriones) [inline src/main/java/org/kairosdb/core/reporting/TagTagger.java:116]: nit `dimensions` -> `tags`? same below

File: src/main/java/org/kairosdb/core/reporting/TagTagger.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+/**
+ * Applies the metric name as a tag.
```
Comment (cwbriones) [inline src/main/java/org/kairosdb/core/reporting/TagTagger.java:31]: This docstring is incorrect

File: src/main/java/org/kairosdb/core/reporting/TagTagger.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                    for (final String tagValue : tags.get(tagName)) {
+                        tagConsumer.accept(tagTargetName, tagValue);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * {@link com.arpnetworking.commons.builder.Builder} implementation for {@link MetricNameTagger}.
```
Comment (cwbriones) [inline src/main/java/org/kairosdb/core/reporting/TagTagger.java:76]: This docstring is incorrect

File: src/main/java/org/kairosdb/core/reporting/Tagger.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+package org.kairosdb.core.reporting;
+
+import com.google.common.collect.SetMultimap;
+
+import java.util.function.Supplier;
+
+public interface Tagger {
+
+    /**
+     * Applies the tags from the available information to the
```
Comment (cwbriones) [inline src/main/java/org/kairosdb/core/reporting/Tagger.java:25]: Can you add another sentence here elaborating on the usage? It isn't immediately clear from the interface what `metricName` is supposed to be used for or if it's expected to sometimes go unused. I expect this is a result of unifying the metric name tagging (`MetricNameTagger`) and mapped tags (`TagTagger`) under the same interface, but I'd rather be explicit in the documentation.

**Review by cwbriones — COMMENTED** [review]
Doc / comment nits.

Overall logic looks good, similar to what was done in CAGG.

File: src/main/java/org/kairosdb/core/reporting/MetricReportingModule.java (hunk 1 above)
Comment (vjkoskela) [reply src/main/java/org/kairosdb/core/reporting/MetricReportingModule.java:115]: Fixed.

File: src/main/java/org/kairosdb/core/reporting/TagTagger.java (hunk 2 above)
Comment (vjkoskela) [reply src/main/java/org/kairosdb/core/reporting/TagTagger.java:116]: Fixed!

File: src/main/java/org/kairosdb/core/reporting/TagTagger.java (hunk 3 above)
Comment (vjkoskela) [reply src/main/java/org/kairosdb/core/reporting/TagTagger.java:31]: Fixed.

File: src/main/java/org/kairosdb/core/reporting/TagTagger.java (hunk 4 above)
Comment (vjkoskela) [reply src/main/java/org/kairosdb/core/reporting/TagTagger.java:76]: Fixed.

File: src/main/java/org/kairosdb/core/reporting/Tagger.java (hunk 5 above)
Comment (vjkoskela) [reply src/main/java/org/kairosdb/core/reporting/Tagger.java:25]: How's that?

**Review by cwbriones — APPROVED** [review]

**Pull request merged**